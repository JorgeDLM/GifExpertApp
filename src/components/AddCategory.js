import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({categorias, setCategorias, pruebas}) =>{

    // Hooks:
    const [inputValue, setInputValue] = useState('')
    const [mostrarAlerta, setMostrarAlerta] = useState({input:true})
    const [categoriaRepetida, setCategoriaRepetida] = useState({categoria: true})

    // Funciones
    const convertirMayusculas = (e) => e.charAt(0) === ' ' ? e.charAt(1).toUpperCase() + e.toLowerCase().slice(1) : e.charAt(0).toUpperCase() + e.toLowerCase().slice(1)

    const handleAgregar = () => {
        // funciona sin que ponga categorias => pero como en las pruebas tengo q verifique q es funcion lo pongo asi:
        setCategorias( categorias => [convertirMayusculas(inputValue), ...categorias]);
        setInputValue('')
    }

    const sinCambio = inputValue !== '' && inputValue !== ' '
    
    // para saber si hay repetidas
    const contieneTextoRepetido = categorias.indexOf(inputValue) === -1
    // para sabir si hay repetido y aun no hay cambios ene l iput
    const sinCambioTextoRepetido = sinCambio && contieneTextoRepetido

    // input en ENTER se activa:
    const handleSubmit = (e) => {
        // esto lo que hace es evitar el funcionamiento por defecto del form:
        e.preventDefault();
        
        // pruebas && console.log('submit activado', inputValue)
        // utilizando el form al picar enter en el input se activa el submit y me evito este codigo:
        // e => (e.keyCode === 13)
    }

    
    const handleChange = (e) => {
        // manejandoDobleEspacio
        setInputValue(e.target.value.replace(/ +(?= )/g,''))
        // pruebas && console.log( 'handleChange llamado')
    }
    
    const agregarCategoria = () => {
        (sinCambioTextoRepetido) && handleAgregar(); 
        (sinCambioTextoRepetido) && setMostrarAlerta(false); 
        (sinCambio && !contieneTextoRepetido) && setCategoriaRepetida(false); 
        (sinCambio && contieneTextoRepetido) && setCategoriaRepetida(true);
        // pruebas && console.log('Se agregaron gifs')
    }

    // styles
    const boton = {backgroundColor: 'white', borderRadius: '10px', paddingBottom: '2px', marginLeft: '5px'}
    const alerta = {fontSize: '12px', paddingTop: '3px',}

    return(
        <>
            {pruebas && <h2>{inputValue}</h2>}
            <form onSubmit={handleSubmit}>
                <input 
                    maxLength={20} 
                    type="text"
                    placeholder="Buscar GIF's" 
                    value={convertirMayusculas(inputValue)} 
                    // ya no es necesario porque esta el form onEnter:
                    // onKeyDown={e => (e.keyCode === 13) && agregarCategoria()} 
                    onChange={handleChange} 
                />
                <button onClick={agregarCategoria}  style={{...boton}}>Agregar</button>
            </form>

           {(!mostrarAlerta && categoriaRepetida) && 
                <div style={{...alerta, color: 'green'}}>
                    Gifs de {categorias[categorias.length -1 ]} agregados con exito!
                </div>
           }

            {!(categoriaRepetida) && 
                <div style={{...alerta, color: 'red'}}>Gifs repetidos.</div>
            }
        </>
    )
}

AddCategory.propTypes = {
    setCategorias: PropTypes.func.isRequired
}


export default AddCategory;