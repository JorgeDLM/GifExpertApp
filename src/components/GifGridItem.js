import React from 'react'
// para importar propTypes rapido: imp tab
import PropTypes from 'prop-types'

export const GifGridItem = ({title, url, id, ultimaCategoria, categorias}) => {
    return (
        <div className={"card"}>
            <img style={{}} src={url} alt={title}/>
            <p style={{fontWeight: 'bold'}}>{title}</p>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}

GifGridItem.defaultValue = {
    title: 'Hola'
}


export default GifGridItem