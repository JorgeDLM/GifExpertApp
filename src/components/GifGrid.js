import React from 'react'
import GifGridItem from './GifGridItem'
// import { getGifs } from '../helpers/getGifs'
import { useFetchGifs } from '../hooks/useFetchGifs'

export const GifGrid = ({categorias, cantidad, ultimaCategoria, pruebas}) => {


    // Cuando esto se ejecuta lo que hace es que hace que se vuelva a ejecutar el getGifs cada que haya un cambio, esto es porque hay un cambio en este nivel en el componente
    // const [count, setCount] = useState(0);

    // const [images, setImages] = useState([])

    // commente esto porque voy a hacer un custom hook, los hooks siempre empiezan con useFuncionAlgo esta en carpeta de hooks
    // useEffect(() => {
    //     getGifs(categorias, cantidad, apiKey).then(img => setImages(img));

    //     // el segundo argumento es un arreglo de dependencias, sirve para parar el que se ejecute mas el getGifs, esta parte es igual al componentDidMount()
    // }, [categorias, cantidad])

    // const state = useFetchGifs()
    // lo imsmo pero desestructurado
    // para darle un nombre especifico a la data se usan : data:images
    const { data:images,loading } = useFetchGifs(categorias, cantidad)
    

    // si esto lo pongo aqui se ejecuta cuando carga
    // getGifs()
    // lo pase arriba a useEffect para que no se ejecute el codigo cada vez que haya un cambio en el componente
    return (
            <>
            {pruebas && <h2>{images.length}</h2>}
                <div style={{textAlign: 'center', marginBottom: '4px', fontSize: '25px', fontWeight: 'bolder'}}>{categorias}</div>
                {loading && <p className="animate__flash">
                    Cargando...
                </p> }
                    <div className="card-grid">
                        {/* <button onClick={() => setCount(count +1)}>Set count +1</button> */}
                        {images.reverse().map(image =>
                            <GifGridItem
                            ultimaCategoria={ultimaCategoria}
                            categorias={categorias}
                            lastItem={[categorias[1]] && true}
                            key={image.id}
                            // esto es como yo lo hacia
                            // images={image}
                            // id={id} etc...
                            // esto es como debe de ser:
                            // mando un spread de la informacion de image como props, cada prop de las imagenes se manda como prop independiente
                            {...image}
                            />
                        )}
                    </div>
            </>
    )
}
