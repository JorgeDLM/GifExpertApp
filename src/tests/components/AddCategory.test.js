// imr tab => import React from 'react'
import React from 'react'
// esto se pone para ver los shortcuts (toBe, toMatch,etc):
import '@testing-library/jest-dom'
import AddCategory from "../../components/AddCategory"
import { shallow } from "enzyme"
import { GifGrid } from '../../components/GifGrid'

describe('<AddCategory/>', () => {
    const categorias = 'Rick n Morty'
    
    // esta funcion vacia se puede reemplazar con una funcion de jest:
    // const funcion = () => {}
    const funcion = jest.fn()
    // los componentes let los defino doble para quye me ponga la ayuda jest
    let wrapper = shallow( <AddCategory categorias={categorias} setCategorias={funcion} pruebas/> )
    let input = wrapper.find('input').at(0)

    // wrapper de gifs que no use del todo:
    let gifs = shallow(<GifGrid categorias={categorias} cantidad={1} pruebas/>)
    
    beforeEach( () => {
        // esto se ocupa para limpiar todas las simulaciones
        jest.clearAllMocks()
        wrapper = shallow( <AddCategory categorias={categorias} setCategorias={funcion} pruebas/> )
        gifs = shallow(<GifGrid categorias={categorias} cantidad={1} pruebas/>)
        input = wrapper.find('input').at(0)
    })
    
    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot()
    })
    test('debe de cambiar el inputValue con el handleChange', () => {
        const valorPrueba = 'Rick n Morty'
        
        // lo siguiente se hace para definir el evento de e ('algo', {evento}):
        // como el evento es e.target.value entonces seria un objeto dentro de un objeto con un valor, se define asi:
        input.simulate('change', { target: { value: valorPrueba } })
        const inputValue = wrapper.find('h2').at(0).text().trim()
        
        expect( inputValue ).toBe( valorPrueba )
    })
    
    test('onClick debe mostrar gifs', () => {
        const valorPrueba = 'rick n morty'
        input.simulate('change', { target: { value: valorPrueba } })
        wrapper.find('button').at(0).simulate('click')
        // queria buscar gifs en otro componente4 pero por alguna razon no se muestran, creo que es porque inicializa en 0 y empieza a incrementar conforme va encontrando gifs
        // const hayGifs = gifs.find('h2').at(0).text().trim()
        // console.log(hayGifs)
        
        expect( funcion ).toHaveBeenCalled()
        
    })
    
    test('onClick no debe mostrar gifs si no hay texto', () => {
        const valorPrueba = ' '
        input.simulate('change', { target: { value: valorPrueba } })

        wrapper.find('button').at(0).simulate('click')
        
        expect( funcion ).not.toHaveBeenCalled()
        
    })
    
    test('onEnter en el input (handleSumit) debe mostrar gifs', () => {
        // el preventDefault lo ejecuto con (){}
        wrapper.find('form').at(0).simulate('submit', { preventDefault(){} })
        // lo que busco es que nunca se haya ejecutado la funcion en AddCategory porque no hay categoria
        // asi se pone la negacion
        expect( funcion ).not.toHaveBeenCalled()
    })
    
    test('debe de llamar el handleAgregar y limpiar el input', () => {
        const valorPrueba = 'rick n morty'
        input.simulate('change', { target: { value: valorPrueba } })
        wrapper.find('button').at(0).simulate('click') 

        // vuelvo a llamar el h2 para ver el valor postClick
        const valorFinalInput = wrapper.find('h2').at(0).text().trim()

        // espero que la funcion se haya ejecutado una vez simulate(click)
        expect( funcion ).toHaveBeenCalledTimes(1)
        expect(valorFinalInput).toBe('')
        // tambien puedo encontrar el valor del input asi:
        expect(wrapper.find('input').prop('value')).toBe('')
        
        // espero que la funcion se haya ejecutado con una funcion en setCategorias
        expect( funcion ).toHaveBeenCalledWith( expect.any(Function) )
        
    })
    
    
    
    
})
