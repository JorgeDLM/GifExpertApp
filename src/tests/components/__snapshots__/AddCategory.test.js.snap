// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCategory/> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h2 />
  <form
    onSubmit={[Function]}
  >
    <input
      maxLength={20}
      onChange={[Function]}
      placeholder="Buscar GIF's"
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": "10px",
          "marginLeft": "5px",
          "paddingBottom": "2px",
        }
      }
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
